lib.php

```diff
@@ -933,6 +740,239 @@ function plagiarism_urkund_format_temp_content($content, $strippretag = false) {
+                $elements == 'urkund_resubmit_on_close') {
+
+                unset($plagiarismelements[$key]);
+            }
+        }
+
+        // First get existing values.
+        if (empty($data->coursemodule)) {
+            debugging("URKUND settings failure - no coursemodule set in form data, URKUND could not be enabled.");
+            return;
+        }
+        $existingelements = $DB->get_records_menu('plagiarism_urkund_config', array('cm' => $data->coursemodule),
+            '', 'name, id');
+        foreach ($plagiarismelements as $element) {
+            // Don't allow changes to receiver address if urkund is disabled.
+            if (empty($data->use_urkund) && $element == 'urkund_receiver') {
+                continue;
+            }
+            $newelement = new stdClass();
+            $newelement->cm = $data->coursemodule;
+            $newelement->name = $element;
+            if (isset($data->$element) && is_array($data->$element)) {
+                $newelement->value = implode(',', $data->$element);
+            } else {
+                $newelement->value = (isset($data->$element) ? $data->$element : 0);
+            }
+            if (isset($existingelements[$element])) {
+                $newelement->id = $existingelements[$element];
+                $DB->update_record('plagiarism_urkund_config', $newelement);
+            } else {
+                $DB->insert_record('plagiarism_urkund_config', $newelement);
+            }
+
+        }
+        // Don't save user preference if this assignment doesn't use Urkund.
+        if (!empty($data->urkund_receiver) && !empty($data->use_urkund)) {
+            set_user_preference('urkund_receiver', trim($data->urkund_receiver));
+        }
+    }
+}
+
+/**
+ * Validate receiver address
+ *
+ * @param stdClass $data
+ * @return array $errors
+ */
+function plagiarism_urkund_coursemodule_validation($data) {
+    $data = $data->get_submitted_data();
+    if (array_key_exists('use_urkund', $data)) {
+        if ($data->use_urkund) {
+            $plugin = new plagiarism_plugin_urkund();
+            $result = $plugin->validate_receiver($data->urkund_receiver);
+            if ($result === 404) {
+                return array('urkund_receiver' => get_string('receivernotvalid', 'plagiarism_urkund'));
+            }
+        }
+    }
+    return array();
+}
+
+/**
+ * Hook to add plagiarism specific settings to a module settings page
+ *
+ * @param moodleform $formwrapper
+ * @param MoodleQuickForm $mform
+ */
+function plagiarism_urkund_coursemodule_standard_elements($formwrapper, $mform) {
+    global $DB, $PAGE, $CFG;
+
+    $plugin = new plagiarism_plugin_urkund();
+    $plagiarismsettings = $plugin->get_settings();
+    if (!$plagiarismsettings) {
+        return;
+    }
+
+    $cmid = null;
+    if ($cm = $formwrapper->get_coursemodule()) {
+        $cmid = $cm->id;
+    }
+    $modulename = "";
+    if (!empty($cm)) {
```

Comment:
This shouldn't be wrapped in a call to $cm - the point of this code is to check if this plugin type should add plagiarism settings. That should happen on both "new activity" and "edit existing activity" actions.